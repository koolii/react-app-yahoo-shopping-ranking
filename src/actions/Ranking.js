import fetchJsonp from 'fetch-jsonp'
import qs from 'qs'
import { replace } from 'react-router-redux'
import { api } from '../yahoo-api.json'

const API_URL = 'https://shopping.yahooapis.jp/ShoppingWebService/V1/json/categoryRanking'
const APP_ID = api

// このアクションの時にstate情報をリセットする
const startRequest = (category) => ({
  type: 'START_REQUEST',
  payload: { category },
})
// APIからのレスポンスを適用させる
const receiveData = (category, error, response) => ({
  type: 'RECEIVE_DATA',
  payload: { category, error, response },
})
const finishRequest = (category) => ({
  type: 'FINISH_REQUEST',
  payload: { category },
})

// redux-thunk
export const fetchRanking = (categoryId) => {
  return async (dispatch, getState) => {
    // カテゴリIDに紐づくstate.shopping.categoriesにアクセス
    // 存在しないcategoryIdでHTTPアクセスしない(存在しないIDだった場合は'/'にリダイレクト)
    const categories = getState().shopping.categories
    const category = categories.find((category) => (category.id === categoryId))
    if (typeof category === 'undefined') {
      dispatch(replace('/'))
      return;
    }

    // 当てはまるcategoryIdが存在したので正常にHTTPアクセスを行なう
    dispatch(startRequest(categoryId))

    const queryString = qs.stringify({
      appid: APP_ID,
      category_id: categoryId,
    })

    try {
      const response = await fetchJsonp(`${API_URL}?${queryString}`)
      const data = await response.json()
      console.log('RECEIVE_DATA')
      dispatch(receiveData(category, null, data))
    } catch (err) {
      dispatch(receiveData(category, err))
    }
    console.log('FINISHED')
    dispatch(finishRequest(category))
  }
}
