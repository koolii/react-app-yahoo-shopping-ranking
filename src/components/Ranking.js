import React from 'react'
import PropTypes from 'prop-types'
import Card, { CardMedia, CardContent, CardActions } from 'material-ui/Card'
import Typography from 'material-ui/Typography'
import Button from 'material-ui/Button'

export default class Ranking extends React.Component {
  componentWillMount() {
    this.props.onMount(this.props.categoryId)
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.categoryId !== nextProps.categoryId) {
      this.props.onUpdate(nextProps.categoryId)
    }
  }

  render() {
    const { category, ranking, error } = this.props

    // categoryIdからAPIへアクセスして情報を取得する
    return (
      <div>
        {/* ランキングのタイトル */}
        <h2>{
          typeof category !== 'undefined' && category.name
          ? `${category.name}'s Ranking`
          : ''
        }</h2>

        {(() => {
          if (error) {
            // show error
            return <p>occured error, please reload this page.</p>
          } else if (typeof ranking === 'undefined') {
            return <p>Now loading...</p>
          } else {
            return (
              <ol>
                {
                  // warning: immediately returns html element
                  ranking.map((item, i) => (
                    <Card
                      key={`ranking-item-${item.code}`}
                      style={{ maxWidth: '500px', margin: '32px auto' }}
                    >
                      <CardMedia
                        image={item.imageUrl}
                        title={`${i + 1}位 ${item.name}`}
                        style={{ height: '200px' }}
                      />
                      <CardContent>
                        <Typography type="title">
                          {`${i + 1}位 ${item.name}`}
                        </Typography>
                      </CardContent>
                      <CardActions>
                        <Button
                          raised
                          color="secondary"
                          fullWidth
                          href={item.url}
                        >Go to product page</Button>
                      </CardActions>
                    </Card>
                  ))
                }
              </ol>
            )
          }
        })()}
      </div>
    )
  }
}

Ranking.propTYpes = {
  categoryId: PropTypes.string,
  onMount: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,

  // category, ranking, errorを追加
  category: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
  }),

  ranking: PropTypes.arrayOf(
    PropTypes.shape({
      code: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      url: PropTypes.string.isRequired,
      imageUrl: PropTypes.string.isRequired,
    })
  ),

  error: PropTypes.bool.isRequired,
}
Ranking.defaultProps = {
  // categoryId=1 is all kinds
  categoryId: '1',
}
