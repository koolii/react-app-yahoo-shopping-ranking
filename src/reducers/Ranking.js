// 何もしないReducer
// exportしてるnoopが
// １つのstateを特定するためのkeyとして動作する
// export const noop = (state = {}) => state;

const getRanking = (response) => {
  console.log('GET RANKING GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG')
  const ranking = []
  const itemLength = response.ResultSet.totalResultsReturned

  for (let index = 0; index < itemLength; index += 1) {
    // const item = response.ResultSet['0'].Result[`${index}`]
    const item = response.ResultSet['0'].Result['0']
    ranking.push({
      code: item.Code,
      name: item.Name,
      url: item.Url,
      imageUrl: item.Image.Medium,
    })
  }
}

const initialState = {
  category: undefined,
  ranking: undefined,
  error: false,
}
const log = (action) => {
  console.log(`NOW ACTION TYPE IS ${JSON.stringify(action)}`)
}

export default (state = initialState, action) => {
  log(action)
  switch (action.type) {
    // このアクションの時にstate情報をリセットする
    case 'START_REQUEST':
      console.log('REDUCER START_REQUEST')
      return {
        category: action.payload.category,
        ranking: undefined,
        error: false,
      }
    case 'RECEIVE_DATA':
      console.log('(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((')
      console.log(JSON.stringify(action.playload.response))
      console.log('(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((')
      return action.error ? { ...state, error: true } : { ...state, ranking: getRanking(action.payload.response) }
    default:
      console.log('OH MY GOSH')
      console.log('OH MY GOSH')
      console.log('OH MY GOSH')
      console.log('OH MY GOSH')
      return state
  }
}