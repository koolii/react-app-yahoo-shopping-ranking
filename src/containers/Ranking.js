import { connect } from 'react-redux'
import Ranking from '../components/Ranking'
import * as actions from '../actions/Ranking'

// Reducerを定義後に実装を行うこと
const mapStateToProps = (state, ownProps) => {
  console.log(JSON.stringify(state))
  return {
    categoryId: ownProps.categoryId,
    // カテゴリ、ランキング情報をRankingコンポーネントに渡す
    category: state.Ranking.category,
    ranking: state.Ranking.ranking,
    error: state.Ranking.error,
  }
}

const mapDispatchToProps = (dispatch) => ({
  onMount(categoryId) {
    dispatch(actions.fetchRanking(categoryId))
  },
  onUpdate(categoryId) {
    dispatch(actions.fetchRanking(categoryId))
  },
})

export default connect(mapStateToProps, mapDispatchToProps)(Ranking)
