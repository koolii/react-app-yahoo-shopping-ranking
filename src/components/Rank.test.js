// ここでは記述量が多くて辛いからスナップショットによるテストだけを行なう
// 雑

import React from 'react'
import renderer from 'react-test-renderer'
import Ranking from './Ranking'

// super rough
test('Component(Ranking)Snapshot', () => {
  const props = {
    category: undefined,
    ranking: [],
    error: false,
    onMount: () => {},
    onUpdate: () => {},
  }
  const result = renderer.create(<Ranking {...props} />).toJSON()

  expect(result).toMatchSnapshot()
})
