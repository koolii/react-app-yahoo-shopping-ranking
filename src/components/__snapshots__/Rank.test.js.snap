// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(Ranking)Snapshot 1`] = `
<div>
  <h2>
    
  </h2>
  <ol />
</div>
`;
